### ISSUE TITLE
Test Results

### ISSUE THREAD

**Issue by dbcallan** [ddsol/speedtest.net#4]
I am running this test on a raspberry pi 2m and my laptop, and the test results differ greatly than what I am seeing from my speedtest.net in a browser.

My browser speedtest gets ~ 40mbps for download and 50 Mbps upload (which is what we have here at the office)

When I run this script from the command line in linux and windows I get ~20Mbps download and ~30Mbps upload (I had to convert from bytes to bits as bytes is not a very useful measurement for these types of tests).

Also, when running the script it sometimes chooses a server over 100 miles away when there is a server less than 20 that it sometimes chooses.  Is there a rhyme or reason to this?

Can you check to see if there is something in this script that is causing weird and unexpected results?

Thanks.

**Comment by ddsol** [comment]
I think most of the discrepancy in speed has to do with speed test counting overhead, such as tcp and ip packet headers, as bandwidth (which it technically is). As far as it choosing a far server: speedtest.net recommends servers based on geoip, which can be inaccurate. My code only "pings" a limited number of servers and uses the one it got the fastest response from. It is entirely possible packets get delayed, even multiple times during the same "ping", which is actually an http request. Therefore the wrong server may be picked. Doing more pings would help, but slow down testing.

To be clear, I think an adjustment factor should be used to bring results in line with Ookla's official numbers. I'm just not sure what this factor ought to be...

**Comment by dbcallan** [comment]
The testing can be super slow, the project I'm working on will run the test once an hour headlessly and shove data into a database.

I guess the problem I have is figuring out what that factor will be to bring the results more in line with speedtest's numbers.  Would you have any advice on where to look for this data so I can do a conversion?

Thanks for your prompt response.

**Comment by ddsol** [comment]
I don't know what the conversion number ought to be, or even if that's the right way to fix this issue. Maybe you can run 2 concurrent tests using the same server to check if multi-stream download is faster than single-stream download. It may be that tcp or http overhead or roundtripping slows the download. I haven't adequately tested to make sure it saturates the pipes. Let me know if you find anything.

**Comment by dbcallan** [comment]
That seemed to make it worse.

Also, are you sure the results are in bytes and not bits?  If it was in bits the results would be a lot closer to what I would expect to be seeing now.

Thanks.

**Comment by ddsol** [comment]
I changed the code a bit. Results are now reported in Mbps (1 million bits per second). They are also adjusted to be in-line with speedtest.net results. See the new Readme (bottom) for more info.

I changed the version to 1.0.0 to reflect the API change.

**Comment by dbcallan** [comment]
Here is the new result I get when I run it here at my house:


{ speeds:
   { download: 7,
     upload: 6,
     originalDownload: 731784,
     originalUpload: 627617 },
  client:
   { ip: '74.70.228.171',
     lat: 42.8594,
     lon: -73.7915,
     isp: 'Time Warner Cable',
     isprating: 2.5,
     rating: 0,
     ispdlavg: 20.556,
     ispulavg: 2.494 },
  server:
   { host: 'speedtest1.inoc.net',
     lat: 42.6525,
     lon: -73.7572,
     location: 'Albany, NY',
     country: 'United States',
     cc: 'US',
     sponsor: 'INOC, LLC',
     distance: 23.18,
     distanceMi: 14.4,
     ping: 84.4,
     id: '5723' } }

My speedtest.net from a browser on the same computer comes up with 52.1 mbps X 5.7 Mbps

The math conversion from bytes to bits is off too. 731784? bytes = 5854272 bits (731784 * 8) which is 5.854272Mbps (5854272 / (1024 * 1024)) or are you doing some extra math to make it closer to speedtest.net?  Following the same math the upload would be
5.020936? Mbps which is close to being correct, although 6 is pretty close as speedtest.net is reporting 5.7 Mbps.

What would make the download speed be off by so much?

Sorry to sound like a pain but I am hoping to clear up this project soon.

Thanks.


________________________________

**Comment by ddsol** [comment]
Bandwidth unit factors are scientific, not nerdic, and use base 10 factors (1k=1000), not base 2 (1k=1024). Therefore, 5854272 bits is 5.8 Mbit, which is actually the number you have there as well...

I do employ a fudge factor, as is indicated in the source as well as in the [readme](https://github.com/ddsol/speedtest.net/blob/master/README.md#considerations). The readme, btw, also has links to why there's a difference. Look at [testmy.net](http://testmy.net/ipb/topic/28902-why-do-my-results-differ-from-speedtestnet-ookla-speed-tests/) for a long discussion. In short, Ookla messes with the results, favoring higher speeds (especially for uploads), while tossing out slower results.

Please make sure to update to the latest version, because it was updated today.

So, to recap, the information you are asking for is already available in the [readme](https://github.com/ddsol/speedtest.net/blob/master/README.md#considerations).

**Comment by ddsol** [comment]
Um oops. i just noticed the round-off issue... I rounded to whole bytes before, but now I round to whole Mbps, which is clearly wrong. Fixing :)

**Comment by ddsol** [comment]
Fixed in 1.0.1
